core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsMetricsManagerImpl.java

```diff
@@ -0,0 +1,400 @@
+
+      DestinationMetrics build(Set<Object> metricsIds, boolean histogramEnabled) {
+         assert syncRequests != null;
+         assert asyncRequests != null;
+         assert timedOutRequests != null;
+
+         IntConsumer bytesSentConsumer;
+         if (histogramEnabled) {
+            assert bytesSentSummary != null;
+            bytesSentConsumer = bytesSentSummary::record;
+         } else {
+            assert bytesSentCounter != null;
+            bytesSentConsumer = bytesSentCounter::increment;
+         }
+
+         return new DestinationMetrics(syncRequests, asyncRequests, timedOutRequests, bytesSentConsumer, metricsIds);
+      }
+   }
+
+   private static class DestinationMetrics {
+      final TimerTracker syncRequests;
+      final CounterTracker asyncRequests;
+      final CounterTracker timedOutRequests;
+      final IntConsumer bytesSent;
+      final Set<Object> metricsIds;
+
+      DestinationMetrics(TimerTracker syncRequests, CounterTracker asyncRequests, CounterTracker timedOutRequests, IntConsumer bytesSent, Set<Object> metricsIds) {
+         this.syncRequests = syncRequests;
+         this.asyncRequests = asyncRequests;
+         this.timedOutRequests = timedOutRequests;
+         this.bytesSent = bytesSent;
+         this.metricsIds = metricsIds;
+      }
+
+      void recordSyncMessage(long durationNanos) {
+         syncRequests.update(Duration.ofNanos(durationNanos));
+      }
+
+      void incrementBytesSent(int size) {
+         bytesSent.accept(size);
+      }
+
+      void incrementAsyncRequests() {
+         asyncRequests.increment();
+      }
+
+      void incrementTimedOutRequests() {
+         timedOutRequests.increment();
+      }
+
+      void unregister(MetricsRegistry registry) {
+         registry.unregisterMetrics(metricsIds);
+      }
+   }
+
+   private static class RequestTrackerImpl implements RequestTracker {
+      private final Address destination;
+      final DestinationMetrics metrics;
+      final TimeService timeService;
+      volatile long sentTimeNanos;
+      @GuardedBy("this")
+      boolean completed;
+
+      RequestTrackerImpl(Address destination, DestinationMetrics metrics, TimeService timeService) {
+         this.destination = destination;
+         this.metrics = metrics;
+         this.timeService = timeService;
+      }
+
+      @Override
+      public final Address destination() {
+         return destination;
+      }
+
+      @Override
+      public synchronized void resetSendTime() {
+         if (completed) {
+            return;
+         }
+         sentTimeNanos = timeService.time();
+      }
+
+      @Override
+      public synchronized void onComplete() {
+         if (completed) {
+            return;
+         }
+         metrics.recordSyncMessage(timeService.timeDuration(sentTimeNanos, TimeUnit.NANOSECONDS));
+         completed = true;
+      }
+
+      @Override
+      public void onSuspect() {
+         // TODO track suspect events in the future?
```

Comment:
If this is something we want to consider outside of this PR, then I think we should track in a Jira and remove the `onSuspect` calls as it's trivial to re-add and currently it's just adding redundant method calls.

Reply:
Second though, I don't think it makes sense to keep track of this. I'm removing the metrics related to a destination when the view changes and that member is not in the view, so this information will be lost anyway.